Студент Жилкин Федор, 344
# Домашняя работа 3: "Решение частичной проблемы собственных значений."
**Задания:**

1. Степенной метод 
2. Метод скалярных произведений

In [1]:
import numpy as np

In [2]:
A = np.array([[1., 2., 1.], [0., 5., 1.], [6., 4, -1.]])
A

array([[ 1.,  2.,  1.],
       [ 0.,  5.,  1.],
       [ 6.,  4., -1.]])

In [3]:
# Необходимая точность
eps = 0.0001

## Степенной метод 

In [54]:
def power_method(A):
    
    # Рандомный вектор
    b = np.random.rand(A.shape[1])
    
    # Максимальное (по модулю) собственное значение 
    real_eigenvalue = max(abs(np.linalg.eigvals(A)))
    
    # Количество итераций
    i = 0
    
    eigenvalue = -1;
    while(abs(eigenvalue - real_eigenvalue) > eps):
        b_new = np.dot(A, b)
        b_new_norm = np.linalg.norm(b_new)
        b = b_new / b_new_norm
        
        # Собственное значение
        Ab = A.dot(b)
        eigenvalue = b.dot(Ab)
        
        i += 1
        
    # Собственный вектор    
    eigenvector = b 
    
    return eigenvalue, eigenvector, i

In [50]:
power_method(A)

(6.078953466706857, array([0.38101743, 0.62851808, 0.67807871]), 14)

**Максимальное собственное значение:** $ \lambda = 6.0789$ 

**Вектор:** 
$v = 
\begin{pmatrix}
 0.3810155 \\
 0.62851587 \\
 0.67808184
\end{pmatrix}
$

**Решение с заданной точностью получено за 14 итераций**

## Метод скалярных произведений

In [55]:
def dot_product_method(A):
    
    # Рандомные вектора X, Y
    X = np.random.rand(A.shape[1])
    Y = np.random.rand(A.shape[1])
    
    # Максимальное (по модулю) собственное значение 
    real_eigenvalue = max(abs(np.linalg.eigvals(A)))
    
    # Количество итераций
    i = 0
    
    eigenvalue = -1;
    while(abs(eigenvalue - real_eigenvalue) > eps):
        X_new = np.dot(A, X)
        Y_new = np.dot(A.transpose(), Y)

        # Собственное значение
        eigenvalue = np.dot(X_new, Y_new)/np.dot(X, Y_new)
        
        X = np.copy(X_new)
        Y = np.copy(Y_new)
        
        i += 1
    
    # Собственный вектор 
    eigenvector = X_new / np.linalg.norm(X_new)
    
    return eigenvalue, eigenvector, i

In [57]:
dot_product_method(A)

(6.078859972324903, array([0.38080501, 0.62826886, 0.67842891]), 7)

**Максимальное собственное значение:** $ \lambda = 6.0788$ 

**Вектор:** 
$v = 
\begin{pmatrix}
 0.38080501 \\
 0.62826886 \\
 0.67842891
\end{pmatrix}
$

**Решение с заданной точностью получено за 7 итераций**